src/blockdev.rs

```diff
@@ -566,39 +597,58 @@ impl SavedPartitions {
             });
         }
 
+        // Create a vector of the partitions
+        let mut partitions = Vec::new();
+
         // read GPT
-        let gpt = match GPT::find_from(disk) {
-            Ok(gpt) => gpt,
+        match GPT::find_from(disk) {
+            Ok(gpt) => {
+                // cross-check GPT sector size
+                Self::verify_gpt_sector_size(&gpt, sector_size)?;
+
+                // save partitions accepted by filters
+                for (i, p) in gpt.iter() {
+                    if Self::matches_filters(i, p, filters) {
+                        partitions.push((i, p.clone()));
+                    }
+                }
+            }
             Err(gptman::Error::InvalidSignature) => {
-                // ensure no indexes are listed to be saved from a MBR disk
-                // we don't need to check for labels since MBR does not support them
+                // no GPT, check for MBR
                 if filters
                     .iter()
                     .any(|f| matches!(f, PartitionFilter::Index(_, _)))
                     && disk_has_mbr(disk).context("checking if disk has an MBR")?
                 {
-                    bail!("saving partitions from an MBR disk is not yet supported");
+                    let mbr = MBR::read_from(disk, u32::try_from(sector_size)?)
+                        .context("reading disk as MBR")?;
+
+                    // cross-check MBR sector size
+                    Self::verify_mbr_sector_size(&mbr, sector_size)?;
+
+                    for (i, p) in mbr.iter() {
+                        if Self::matches_filters_mbr(i.try_into().unwrap(), p, filters) {
```

Comment:
The `unwrap` is inconsistent with the `try_into().context()?` below.  It seems very unlikely that even a crafted partition table would overflow a u32; let's just be consistent in how we handle the conversion.

---

This is starting to feel too deeply nested, and it seems a bit uncomfortable that `matches_filters_mbr` is almost the same as `matches_filters`.  Would it make sense to swap the order of translation and filtering, by translating the entire MBR to GPT (in a helper function) and then using the original flow?  We'd probably need to add an argument to `matches_filters` to disable matching by label (so `*` doesn't match translated MBR partitions) but that's still less code duplication.

Or actually, maybe it's okay for `*` to match MBR partitions.